### DISCUSSION TITLE
atomic three function with ipopt

### DISCUSSION THREAD

**Discussion by EmmanuelSchneider** [coin-or/CppAD#152] [Q&A] [answered]
Hello,

I'm trying to use an atomic three function with the cppad interface for ipopt.
I have implemented all the callbacks members except the reverse ans rev_depend functions because I use the "forward" and "sparse" option for ipopt.
The compiled code produce a memory error just after Ipopt is called (I see the first lines of Ipopt solver that give us the number of non zeros ...)

Could you tell me if this problem can appear if we don't implemented all the callbacks functions ?

Regards.

Emmanuel.

**Comment by EmmanuelSchneider** [comment, marked as answer]
I had a look at the cppad/ipopt/solve.hpp and cppad/ipopt/solve_callback.hpp files 
and I confirm that all the callbacks methods are exploited.
Since I implemented the reverse and rev_depend callbacks methods the compiled code works, there are no more memory error.

Many thanks to Bradley and all the contributors.

Emmanuel.

**Comment by bradbell** [comment]
@EmmanuelSchneider Is there a reason you prefer the atomic_three interface to the atomic_four interface ?

**Comment by EmmanuelSchneider** [comment]
Hello Bradley,

I use currently the CppAD stable/20190200 under a 20.04 Xubuntu distribution.
I tried to install the latest version of CppAD but I had some troubles.
Because in the 20.04 ubuntu the official package of Ipopt is 3.11.9, I get a compilation error when I compile the ipopt_get_started : The compiler tell us that the "coin-or/IpIpoptApplication.hpp" file doesn't exist. I seen effectively that the headers files of the 3.11.9 Ipopt version are in a "coin" directory. In a more recent Ipopt version like the 3.14, the headers are in "coin-or" directory. I began to compile/install Ipopt 3.14 + MUMPS but it seems not to be so easy.

Are there a big difference between the three and four atomic interface ?

Regards.

Emmanuel.

**Comment by bradbell** [comment]
What is the value of cppad_postfix in your cmake command ?
https://coin-or.github.io/CppAD/doc/cmake.htm#cppad_postfix

The big difference between atomic_three versus atomic_four is the use of call_id
https://coin-or.github.io/CppAD/doc/atomic_four_call.htm#call_id
for extra information that is related to the specific call to an atomic function.

This avoids having to encode this information in the function argument which gets confusing and is limited. For example call_id could identify the sparsity pattern for a sparse matrix.

**Comment by EmmanuelSchneider** [comment]
Hello Bradley,

The cppad_posfix is empty (I didn't use it).
Because it's a lot of months I had installed cppad, I tried to install the master branch again to be sure what I tell you

My commands in the folder where I cloned the git cppad project were

> mkdir build_2022_ipopt
> cmake -S . -B build_2022_ipopt/ -DCMAKE_BUILD_TYPE=Release -Dcppad_prefix=/home/CppAD/install_2022_ipopt -Dinclude_ipopt=true

==> The configuration seems ok, in particular it tells us that it found ipopt 3.11.9

> cmake --build build_2022_ipopt/
 
  Scanning dependencies of target cppad_lib
  [  0%] Building CXX object cppad_lib/CMakeFiles/cppad_lib.dir/cpp_graph_op.cpp.o
  [  0%] Building CXX object cppad_lib/CMakeFiles/cppad_lib.dir/cppad_colpack.cpp.o
  [ 50%] Building CXX object cppad_lib/CMakeFiles/cppad_lib.dir/json_lexer.cpp.o
  [ 50%] Building CXX object cppad_lib/CMakeFiles/cppad_lib.dir/json_parser.cpp.o
  [ 50%] Building CXX object cppad_lib/CMakeFiles/cppad_lib.dir/json_writer.cpp.o
  [ 50%] Linking CXX shared library libcppad_lib.so
  [ 50%] Built target cppad_lib
  Scanning dependencies of target cppad_ipopt
  [ 50%] Building CXX object cppad_ipopt/src/CMakeFiles/cppad_ipopt.dir/cppad_ipopt_nlp.cpp.o
  In file included from /home/CppAD/include/cppad/ipopt/solve.hpp:400,
                   from /home/CppAD/include/cppad/cppad.hpp:73,
                   from /home/CppAD/cppad_ipopt/src/cppad_ipopt_nlp.hpp:576,
                   from /home/CppAD/cppad_ipopt/src/cppad_ipopt_nlp.cpp:14:
  /home/CppAD/include/cppad/ipopt/solve_callback.hpp:16:11: fatal error: coin-or/IpIpoptApplication.hpp: Aucun fichier ou dossier de ce type
     16 | # include <coin-or/IpIpoptApplication.hpp>
        |           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  compilation terminated.
  make[2]: *** [cppad_ipopt/src/CMakeFiles/cppad_ipopt.dir/build.make:63 : cppad_ipopt/src/CMakeFiles/cppad_ipopt.dir/cppad_ipopt_nlp.cpp.o] Erreur 1
  make[1]: *** [CMakeFiles/Makefile2:3115 : cppad_ipopt/src/CMakeFiles/cppad_ipopt.dir/all] Erreur 2
  make: *** [Makefile:130 : all] Erreur 2

==> Same error than the last time : The compiler tell us that the "coin-or/IpIpoptApplication.hpp" file doesn't exist

I hope this explain you better the problem

Regards.

Emmanuel.

**Comment by bradbell** [comment]
What do you get when you execute
```
pkg-config ipopt --cflags
```

**Comment by EmmanuelSchneider** [comment]
I get -DHAVE_CSTDDEF -I/usr/include/coin

**Comment by bradbell** [comment]
That should work.  Try including the following to your cmake command line:
   -D CMAKE_VERBOSE_MAKEFILE=true
and then report back make output section of the build that reports the error.

**Comment by EmmanuelSchneider** [comment]
When you say that this should work, you mean with the 3.11.9 ipopt version ?
 
After I add the verbose, I get this end of the build message

Scanning dependencies of target cppad_ipopt
make[2] : on quitte le répertoire « /home/CppAD/build_2022_ipopt »
/usr/bin/make -f cppad_ipopt/src/CMakeFiles/cppad_ipopt.dir/build.make cppad_ipopt/src/CMakeFiles/cppad_ipopt.dir/build
make[2] : on entre dans le répertoire « /home/CppAD/build_2022_ipopt »
[ 50%] Building CXX object cppad_ipopt/src/CMakeFiles/cppad_ipopt.dir/cppad_ipopt_nlp.cpp.o
cd /home/CppAD/build_2022_ipopt/cppad_ipopt/src && /usr/bin/c++   -I/home/CppAD/include -isystem /usr/include/coin  -O3 -DNDEBUG   -o CMakeFiles/cppad_ipopt.dir/cppad_ipopt_nlp.cpp.o -c /home/CppAD/cppad_ipopt/src/cppad_ipopt_nlp.cpp
In file included from /home/CppAD/include/cppad/ipopt/solve.hpp:400,
                 from /home/CppAD/include/cppad/cppad.hpp:73,
                 from /home/CppAD/cppad_ipopt/src/cppad_ipopt_nlp.hpp:576,
                 from /home/CppAD/cppad_ipopt/src/cppad_ipopt_nlp.cpp:14:
/home/CppAD/include/cppad/ipopt/solve_callback.hpp:16:11: fatal error: coin-or/IpIpoptApplication.hpp: Aucun fichier ou dossier de ce type
   16 | # include <coin-or/IpIpoptApplication.hpp>
      |           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
make[2]: *** [cppad_ipopt/src/CMakeFiles/cppad_ipopt.dir/build.make:66 : cppad_ipopt/src/CMakeFiles/cppad_ipopt.dir/cppad_ipopt_nlp.cpp.o] Erreur 1
make[2] : on quitte le répertoire « /home/CppAD/build_2022_ipopt »
make[1]: *** [CMakeFiles/Makefile2:3118 : cppad_ipopt/src/CMakeFiles/cppad_ipopt.dir/all] Erreur 2
make[1] : on quitte le répertoire « /home/CppAD/build_2022_ipopt »
make: *** [Makefile:133 : all] Erreur 2

**Comment by bradbell** [comment]
What exactly is your cmake command; see
https://coin-or.github.io/CppAD/doc/cmake.htm#CMake%20Command
